File: book/lib/services/kaggle_service.dart
import 'api_client.dart';

// Models
class KaggleDataset {
  final String ref;
  final String title;
  final String size;
  final String lastUpdated;
  final int downloadCount;
  final int voteCount;
  final double usabilityRating;

  KaggleDataset({
    required this.ref,
    required this.title,
    required this.size,
    required this.lastUpdated,
    required this.downloadCount,
    required this.voteCount,
    required this.usabilityRating,
  });

  factory KaggleDataset.fromJson(Map<String, dynamic> json) {
    return KaggleDataset(
      ref: json['ref'] ?? '',
      title: json['title'] ?? '',
      size: json['size'] ?? '0',
      lastUpdated: json['lastUpdated'] ?? '',
      downloadCount: json['downloadCount'] ?? 0,
      voteCount: json['voteCount'] ?? 0,
      usabilityRating: (json['usabilityRating'] ?? 0).toDouble(),
    );
  }

  String get owner => ref.split('/').first;
  String get name => ref.split('/').last;
}

class KaggleCompetition {
  final String ref;
  final String title;
  final String deadline;
  final String category;
  final String reward;
  final int teamCount;
  final bool userHasEntered;

  KaggleCompetition({
    required this.ref,
    required this.title,
    required this.deadline,
    required this.category,
    required this.reward,
    required this.teamCount,
    required this.userHasEntered,
  });

  factory KaggleCompetition.fromJson(Map<String, dynamic> json) {
    return KaggleCompetition(
      ref: json['ref'] ?? '',
      title: json['title'] ?? '',
      deadline: json['deadline'] ?? '',
      category: json['category'] ?? '',
      reward: json['reward'] ?? '',
      teamCount: json['teamCount'] ?? 0,
      userHasEntered: json['userHasEntered'] ?? false,
    );
  }
}

class KaggleKernel {
  final String ref;
  final String title;
  final String author;
  final String lastRunTime;
  final int totalVotes;
  final String language;

  KaggleKernel({
    required this.ref,
    required this.title,
    required this.author,
    required this.lastRunTime,
    required this.totalVotes,
    required this.language,
  });

  factory KaggleKernel.fromJson(Map<String, dynamic> json) {
    return KaggleKernel(
      ref: json['ref'] ?? '',
      title: json['title'] ?? '',
      author: json['author'] ?? '',
      lastRunTime: json['lastRunTime'] ?? '',
      totalVotes: json['totalVotes'] ?? 0,
      language: json['language'] ?? '',
    );
  }
}

class KaggleFile {
  final String name;
  final String size;
  final String? description;
  final String? creationDate;

  KaggleFile({
    required this.name,
    required this.size,
    this.description,
    this.creationDate,
  });

  factory KaggleFile.fromJson(Map<String, dynamic> json) {
    return KaggleFile(
      name: json['name'] ?? '',
      size: json['size'] ?? '0',
      description: json['description'],
      creationDate: json['creationDate'],
    );
  }
}

class KaggleSubmission {
  final String fileName;
  final String date;
  final String description;
  final String status;
  final String publicScore;
  final String privateScore;

  KaggleSubmission({
    required this.fileName,
    required this.date,
    required this.description,
    required this.status,
    required this.publicScore,
    required this.privateScore,
  });

  factory KaggleSubmission.fromJson(Map<String, dynamic> json) {
    return KaggleSubmission(
      fileName: json['fileName'] ?? '',
      date: json['date'] ?? '',
      description: json['description'] ?? '',
      status: json['status'] ?? '',
      publicScore: json['publicScore'] ?? '',
      privateScore: json['privateScore'] ?? '',
    );
  }
}

class KaggleStatus {
  final bool configured;
  final String? username;
  final String message;

  KaggleStatus({
    required this.configured,
    this.username,
    required this.message,
  });

  factory KaggleStatus.fromJson(Map<String, dynamic> json) {
    return KaggleStatus(
      configured: json['configured'] ?? false,
      username: json['username'],
      message: json['message'] ?? '',
    );
  }
}

// Service
class KaggleService {
  final ApiClient _api;

  KaggleService({ApiClient? api}) : _api = api ?? apiClient;

  // ============================================================================
  // STATUS & CREDENTIALS
  // ============================================================================

  Future<KaggleStatus> getStatus() async {
    try {
      final response = await _api.get('/api/kaggle/status');
      return KaggleStatus.fromJson(response);
    } catch (e) {
      return KaggleStatus(configured: false, message: 'Error: $e');
    }
  }

  Future<bool> setCredentials(String username, String key) async {
    try {
      await _api.post('/api/kaggle/credentials', {
        'username': username,
        'key': key,
      });
      return true;
    } catch (e) {
      return false;
    }
  }

  // ============================================================================
  // DATASETS
  // ============================================================================

  Future<List<KaggleDataset>> listDatasets({
    String sortBy = 'hottest',
    int page = 1,
    int pageSize = 20,
  }) async {
    try {
      final response = await _api.get(
        '/api/kaggle/datasets/list?sort_by=$sortBy&page=$page&page_size=$pageSize',
      );
      final datasets = response['datasets'] as List<dynamic>? ?? [];
      return datasets.map((d) => KaggleDataset.fromJson(d)).toList();
    } catch (e) {
      return [];
    }
  }

  Future<List<KaggleDataset>> searchDatasets(
    String query, {
    int page = 1,
    int pageSize = 20,
  }) async {
    try {
      final response = await _api.get(
        '/api/kaggle/datasets/search?query=$query&page=$page&page_size=$pageSize',
      );
      final datasets = response['datasets'] as List<dynamic>? ?? [];
      return datasets.map((d) => KaggleDataset.fromJson(d)).toList();
    } catch (e) {
      return [];
    }
  }

  Future<Map<String, dynamic>> downloadDataset(
    String datasetRef, {
    String? path,
    bool unzip = true,
  }) async {
    try {
      final slug = _extractSlug(datasetRef);
      final response = await _api.post('/api/kaggle/datasets/download', {
        'dataset': slug,
        if (path != null) 'path': path,
        'unzip': unzip,
      });
      return response;
    } catch (e) {
      return {'success': false, 'message': 'Error: $e'};
    }
  }

  Future<List<KaggleFile>> getDatasetFiles(String owner, String datasetName) async {
    try {
      final response = await _api.get('/api/kaggle/datasets/$owner/$datasetName/files');
      final files = response['files'] as List<dynamic>? ?? [];
      return files.map((f) => KaggleFile.fromJson(f)).toList();
    } catch (e) {
      return [];
    }
  }

  // ============================================================================
  // COMPETITIONS
  // ============================================================================

  Future<List<KaggleCompetition>> listCompetitions({
    String category = 'all',
    String sortBy = 'latestDeadline',
    int page = 1,
  }) async {
    try {
      final response = await _api.get(
        '/api/kaggle/competitions/list?category=$category&sort_by=$sortBy&page=$page',
      );
      final competitions = response['competitions'] as List<dynamic>? ?? [];
      return competitions.map((c) => KaggleCompetition.fromJson(c)).toList();
    } catch (e) {
      return [];
    }
  }

  Future<List<KaggleFile>> getCompetitionFiles(String competition) async {
    try {
      final slug = _extractSlug(competition);
      final response = await _api.get('/api/kaggle/competitions/$slug/files');
      final files = response['files'] as List<dynamic>? ?? [];
      return files.map((f) => KaggleFile.fromJson(f)).toList();
    } catch (e) {
      return [];
    }
  }

  Future<Map<String, dynamic>> downloadCompetitionData(
    String competition, {
    String? path,
    String? file,
  }) async {
    try {
      final slug = _extractSlug(competition);
      final response = await _api.post('/api/kaggle/competitions/download', {
        'competition': slug,
        if (path != null) 'path': path,
        if (file != null) 'file': file,
      });
      return response;
    } catch (e) {
      return {'success': false, 'message': 'Error: $e'};
    }
  }

  Future<List<KaggleSubmission>> getSubmissions(String competition) async {
    try {
      // Extract slug from full URL if needed
      final slug = _extractSlug(competition);
      final response = await _api.get('/api/kaggle/competitions/$slug/submissions');
      final submissions = response['submissions'] as List<dynamic>? ?? [];
      return submissions.map((s) => KaggleSubmission.fromJson(s)).toList();
    } catch (e) {
      return [];
    }
  }

  /// Extract slug from Kaggle URL or return as-is if already a slug
  String _extractSlug(String refOrUrl) {
    // If it's a full URL like https://www.kaggle.com/competitions/name or https://www.kaggle.com/datasets/user/name
    if (refOrUrl.startsWith('http')) {
      final uri = Uri.tryParse(refOrUrl);
      if (uri != null && uri.pathSegments.isNotEmpty) {
        // For competitions: /competitions/name -> name
        // For datasets: /datasets/user/name -> user/name
        // For kernels: /code/user/name -> user/name
        final segments = uri.pathSegments;
        if (segments.length >= 2) {
          if (segments[0] == 'competitions') {
            return segments[1];
          } else if (segments[0] == 'datasets' || segments[0] == 'code') {
            return segments.skip(1).join('/');
          }
        }
      }
    }
    return refOrUrl;
  }

  // ============================================================================
  // KERNELS / NOTEBOOKS
  // ============================================================================

  Future<List<KaggleKernel>> listKernels({
    String sortBy = 'hotness',
    int page = 1,
    int pageSize = 20,
  }) async {
    try {
      final response = await _api.get(
        '/api/kaggle/kernels/list?sort_by=$sortBy&page=$page&page_size=$pageSize',
      );
      final kernels = response['kernels'] as List<dynamic>? ?? [];
      return kernels.map((k) => KaggleKernel.fromJson(k)).toList();
    } catch (e) {
      return [];
    }
  }

  Future<List<KaggleKernel>> searchKernels(
    String query, {
    int page = 1,
    int pageSize = 20,
  }) async {
    try {
      final response = await _api.get(
        '/api/kaggle/kernels/search?query=$query&page=$page&page_size=$pageSize',
      );
      final kernels = response['kernels'] as List<dynamic>? ?? [];
      return kernels.map((k) => KaggleKernel.fromJson(k)).toList();
    } catch (e) {
      return [];
    }
  }

  Future<Map<String, dynamic>> pullKernel(String kernelRef, {String? path}) async {
    try {
      final slug = _extractSlug(kernelRef);
      final response = await _api.post('/api/kaggle/kernels/pull', {
        'kernel_ref': slug,
        if (path != null) 'path': path,
      });
      return response;
    } catch (e) {
      return {'success': false, 'message': 'Error: $e'};
    }
  }
}

final kaggleService = KaggleService();

File: notebook/src/services/kaggleService.ts
// Kaggle Service - Integration with Kaggle API

import apiClient from "./apiClient";

export interface KaggleDataset {
  ref: string;
  title: string;
  owner: string;
  name: string;
  subtitle?: string;
  description?: string;
  downloadCount: number;
  voteCount: number;
  size: string;
  lastUpdated: Date;
  license?: string;
  tags?: string[];
}

export interface KaggleCompetition {
  ref: string;
  title: string;
  url: string;
  description: string;
  category: string;
  reward: string;
  teamCount: number;
  deadline: Date;
  userHasEntered: boolean;
  organizationName?: string;
  evaluationMetric?: string;
}

export interface KaggleKernel {
  ref: string;
  title: string;
  author: string;
  language: string;
  kernelType: string;
  totalVotes: number;
  lastRunTime: Date;
}

export interface KaggleSubmission {
  ref: string;
  fileName: string;
  date: Date;
  status: string;
  publicScore?: number;
  privateScore?: number;
}

interface KaggleDatasetResponse {
  ref: string;
  title: string;
  owner?: string;
  name?: string;
  subtitle?: string;
  description?: string;
  download_count?: number;
  downloadCount?: number;
  vote_count?: number;
  voteCount?: number;
  size: string;
  last_updated?: string;
  lastUpdated?: string;
  license?: string;
  tags?: string[];
  usabilityRating?: number;
}

interface KaggleCompetitionResponse {
  ref: string;
  title: string;
  url?: string;
  description?: string;
  category: string;
  reward: string;
  team_count?: number;
  teamCount?: number;
  deadline: string;
  user_has_entered?: boolean;
  userHasEntered?: boolean;
  organization_name?: string;
  evaluation_metric?: string;
}

class KaggleService {
  private parseDataset(data: KaggleDatasetResponse): KaggleDataset {
    // Parse owner/name from ref if not provided
    const refParts = data.ref?.split("/") || [];
    const owner = data.owner || refParts[0] || "";
    const name = data.name || refParts[1] || "";

    return {
      ref: data.ref,
      title: data.title,
      owner,
      name,
      subtitle: data.subtitle,
      description: data.description,
      downloadCount: data.download_count ?? data.downloadCount ?? 0,
      voteCount: data.vote_count ?? data.voteCount ?? 0,
      size: data.size,
      lastUpdated: new Date(data.last_updated || data.lastUpdated || Date.now()),
      license: data.license,
      tags: data.tags,
    };
  }

  private parseCompetition(data: KaggleCompetitionResponse): KaggleCompetition {
    return {
      ref: data.ref,
      title: data.title,
      url: data.url || `https://www.kaggle.com/c/${data.ref}`,
      description: data.description || "",
      category: data.category,
      reward: data.reward,
      teamCount: data.team_count ?? data.teamCount ?? 0,
      deadline: new Date(data.deadline),
      userHasEntered: data.user_has_entered ?? data.userHasEntered ?? false,
      organizationName: data.organization_name,
      evaluationMetric: data.evaluation_metric,
    };
  }

  // Check if Kaggle is configured
  async isConfigured(): Promise<boolean> {
    try {
      const response = await apiClient.get<{ configured: boolean }>("/api/kaggle/status");
      return response.configured;
    } catch {
      return false;
    }
  }

  // Get current user info
  async getUser(): Promise<{ username: string; displayName: string; tier: string } | null> {
    try {
      return await apiClient.get("/api/kaggle/user");
    } catch {
      return null;
    }
  }

  // Datasets
  async searchDatasets(query: string, page: number = 1): Promise<KaggleDataset[]> {
    const response = await apiClient.get<KaggleDatasetResponse[]>(
      `/api/kaggle/datasets/search?query=${encodeURIComponent(query)}&page=${page}`
    );
    return response.map((d) => this.parseDataset(d));
  }

  async listDatasets(page: number = 1, sortBy: string = "hottest"): Promise<KaggleDataset[]> {
    const response = await apiClient.get<{ datasets: KaggleDatasetResponse[]; total: number }>(
      `/api/kaggle/datasets/list?page=${page}&sort_by=${sortBy}`
    );
    return response.datasets.map((d) => this.parseDataset(d));
  }

  async getDataset(owner: string, name: string): Promise<KaggleDataset> {
    const response = await apiClient.get<KaggleDatasetResponse>(
      `/api/kaggle/datasets/${owner}/${name}`
    );
    return this.parseDataset(response);
  }

  async downloadDataset(ref: string, destinationPath: string): Promise<{ success: boolean; path: string }> {
    return apiClient.post("/api/kaggle/datasets/download", {
      ref,
      destination_path: destinationPath,
    });
  }

  // Competitions
  async listCompetitions(page: number = 1, category?: string): Promise<KaggleCompetition[]> {
    const url = category
      ? `/api/kaggle/competitions/list?page=${page}&category=${category}`
      : `/api/kaggle/competitions/list?page=${page}`;
    const response = await apiClient.get<{ competitions: KaggleCompetitionResponse[]; total: number }>(url);
    return response.competitions.map((c) => this.parseCompetition(c));
  }

  async searchCompetitions(query: string): Promise<KaggleCompetition[]> {
    const response = await apiClient.get<KaggleCompetitionResponse[]>(
      `/api/kaggle/competitions/search?query=${encodeURIComponent(query)}`
    );
    return response.map((c) => this.parseCompetition(c));
  }

  async getCompetition(ref: string): Promise<KaggleCompetition> {
    const response = await apiClient.get<KaggleCompetitionResponse>(
      `/api/kaggle/competitions/${ref}`
    );
    return this.parseCompetition(response);
  }

  async downloadCompetitionData(ref: string, destinationPath: string): Promise<{ success: boolean; path: string }> {
    return apiClient.post("/api/kaggle/competitions/download", {
      ref,
      destination_path: destinationPath,
    });
  }

  async submitToCompetition(ref: string, filePath: string, message?: string): Promise<KaggleSubmission> {
    const response = await apiClient.post<{
      ref: string;
      file_name: string;
      date: string;
      status: string;
      public_score?: number;
      private_score?: number;
    }>("/api/kaggle/competitions/submit", {
      ref,
      file_path: filePath,
      message,
    });

    return {
      ref: response.ref,
      fileName: response.file_name,
      date: new Date(response.date),
      status: response.status,
      publicScore: response.public_score,
      privateScore: response.private_score,
    };
  }

  async getSubmissions(competitionRef: string): Promise<KaggleSubmission[]> {
    const response = await apiClient.get<
      Array<{
        ref: string;
        file_name: string;
        date: string;
        status: string;
        public_score?: number;
        private_score?: number;
      }>
    >(`/api/kaggle/competitions/${competitionRef}/submissions`);

    return response.map((s) => ({
      ref: s.ref,
      fileName: s.file_name,
      date: new Date(s.date),
      status: s.status,
      publicScore: s.public_score,
      privateScore: s.private_score,
    }));
  }

  // Kernels/Notebooks
  async listKernels(page: number = 1, language?: string): Promise<KaggleKernel[]> {
    const url = language
      ? `/api/kaggle/kernels/list?page=${page}&language=${language}`
      : `/api/kaggle/kernels/list?page=${page}`;

    const response = await apiClient.get<{
      kernels: Array<{
        ref: string;
        title: string;
        author: string;
        language: string;
        lastRunTime: string;
        totalVotes: number;
      }>;
      total: number;
    }>(url);

    return response.kernels.map((k) => ({
      ref: k.ref,
      title: k.title,
      author: k.author,
      language: k.language,
      kernelType: "notebook",
      totalVotes: k.totalVotes,
      lastRunTime: new Date(k.lastRunTime),
    }));
  }

  async pullKernel(ref: string, destinationPath: string): Promise<{ success: boolean; path: string }> {
    return apiClient.post("/api/kaggle/kernels/pull", {
      ref,
      destination_path: destinationPath,
    });
  }
}

export const kaggleService = new KaggleService();
export default kaggleService;
